modules/openideal_user/src/Plugin/Block/OpenidealUserPersonalActivityStream.php

```diff
@@ -0,0 +1,174 @@
+<?php
+
+namespace Drupal\openideal_user\Plugin\Block;
+
+use Drupal\Core\Block\BlockBase;
+use Drupal\Core\Database\Connection;
+use Drupal\Core\Entity\EntityTypeManager;
+use Drupal\Core\Plugin\ContainerFactoryPluginInterface;
+use Drupal\Core\Session\AccountProxy;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+/**
+ * Provides a 'OpenidealUserPersonalActivityStream' block.
+ *
+ * @Block(
+ *  id = "openideal_personal_activity_stream_block",
+ *  admin_label = @Translation("Personal activity stream"),
+ * )
+ */
+class OpenidealUserPersonalActivityStream extends BlockBase implements ContainerFactoryPluginInterface {
+
+  /**
+   * Limit of activity entities in list.
+   *
+   * @Todo: Move to config form?
+   */
+  const LIMIT = 10;
+
+  /**
+   * Block manager.
+   *
+   * @var \Drupal\Core\Database\Connection
+   */
+  protected $database;
+
+  /**
+   * Entity type manager.
+   *
+   * @var \Drupal\Core\Entity\EntityTypeManager
+   */
+  protected $entityTypeManager;
+
+  /**
+   * Current user.
+   *
+   * @var \Drupal\Core\Session\AccountProxy
+   */
+  protected $currentUser;
+
+  /**
+   * {@inheritDoc}
+   */
+  public function __construct(
+    array $configuration,
+    $plugin_id,
+    $plugin_definition,
+    Connection $database,
+    EntityTypeManager $entityTypeManager,
+    AccountProxy $currentUser
+  ) {
+    parent::__construct($configuration, $plugin_id, $plugin_definition);
+    $this->database = $database;
+    $this->entityTypeManager = $entityTypeManager;
+    $this->currentUser = $currentUser;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition) {
+    return new static(
+      $configuration,
+      $plugin_id,
+      $plugin_definition,
+      $container->get('database'),
+      $container->get('entity_type.manager'),
+      $container->get('current_user')
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function build() {
+    $build['#theme'] = 'openideal_user_activity_list';
+    $entities = $this->getContent();
+    if (empty($entities)) {
+      return $build;
+    }
+    $view_builder = $this->entityTypeManager->getViewBuilder('message');
+    $items = [];
+    /** @var \Drupal\message\Entity\Message $entity */
+    foreach ($entities as $entity) {
+      $items[] = $view_builder->view($entity);
+    }
+
+    $build['#list'] = $items;
+    $build['#cache']['tags'][] = 'message_list';
+    return $build;
+  }
+
+  /**
+   * Fetch activity stream message ids.
+   *
+   * @Todo: Find the way to check permissions when the private flavours module is enabled.
+   * Because of "node_grants" were removed from the Group module.
+   *
+   * @return array|null
+   *   Fetched entities.
+   */
+  protected function fetchContent() {
```

Comment:
1) Can we avoid duplicated code?
2) Can we avoid variables like `second_query`, `third_query`?

Reply:
2. Done.